CHANGELOG.md

```diff
@@ -6,6 +6,258 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+        Config: tpmodels.ProviderConfig{
+            Clients: []tpmodels.ProviderClientConfig{
+                {
+                    ClientType: "web",
+                    ClientID:     "clientid1",
+                    ClientSecret: "...",
+                },
+                {
+                    ClientType: "android",
+                    ClientID:     "clientid2",
+                    ClientSecret: "...",
+                },
+            },
+        },
+    },
+}
+```
+
+3. Change in the implementation of the custom providers
+
+* All the configuration is part of the `ProviderInput`
+* To update `AuthorisationRedirect.Params` dynamically, GetAuthorisationRedirectURL must be overridden
+* To provide implementation for `GetProfileInfo`,
+  * either use `UserInfoEndpoint`, `UserInfoEndpointQueryParams` and `UserInfoMap` to fetch the user info from the provider
+  * or, specify custom implementation in an override for `GetUserInfo`
+
+Before:
+
+```go
+tpmodels.TypeProvider{
+    ID: "custom",
+    Get: func(redirectURI, authCodeFromRequest *string, userContext supertokens.UserContext) tpmodels.TypeProviderGetResponse {
+        return tpmodels.TypeProviderGetResponse{
+            AccessTokenAPI: "...",
+            AuthorisationRedirect: tpmodels.AuthorisationRedirect{
+                URL:    "...",
+                Params: map[string]interface{}{},
+            },
+            GetClientId: func(userContext supertokens.UserContext) string {
+                return "..."
+            },
+            GetRedirectURI: func(userContext supertokens.UserContext) (string, error) {
+                return "...", nil
+            },
+            GetProfileInfo: func(authCodeResponse interface{}, userContext supertokens.UserContext) (tpmodels.UserInfo, error) {
+                return tpmodels.UserInfo{
+                    ID: "...",
+                    Email: &tpmodels.EmailStruct{
+                        ID:         "...",
+                        IsVerified: true,
+                    },
+                }, nil
+            },
+        }
+    },
+}
+```
+
+After:
+
+```go
+tpmodels.ProviderInput{
+    ThirdPartyID: "custom",
+    Config: tpmodels.ProviderConfig{
+        Clients: []tpmodels.ProviderClientConfig{
+            {
+                ClientID:     "...",
+                ClientSecret: "...",
+            },
+        },
+        AuthorizationEndpoint:            "...",
+        AuthorizationEndpointQueryParams: map[string]interface{}{},
+        TokenEndpoint:                    "...",
+        TokenEndpointBodyParams:          map[string]interface{}{},
+        UserInfoEndpoint:                 "...",
+        UserInfoEndpointQueryParams:      map[string]interface{}{},
+        UserInfoMap: tpmodels.TypeUserInfoMap{
+            FromUserInfoAPI: struct {
+                UserId        string "json:\"userId\""
+                Email         string "json:\"email\""
+                EmailVerified string "json:\"emailVerified\""
```

Comment:
Why is it required to give `"json:`..`"` and not just `"userId"`

Reply:
Will create a new type for this so that it simplifies the user input. Have updated the changelog according to that.

Reply:
why?? can't it just use the same as the other map?

Reply:
We discussed this on call. Created a new type to avoid inline struct.